connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java

```diff
@@ -0,0 +1,822 @@
+                        if (adjacentBlockId != BlockStateValues.JAVA_AIR_ID && BlockStateValues.isBlockAttached(blockId, adjacentBlockId) && canMoveBlock(adjacentBlockId, false)) {
+                            // If it is another slime/honey block we need to check its adjacent blocks
+                            if (BlockStateValues.isBlockSticky(adjacentBlockId)) {
+                                blocksToCheck.add(adjacentPos);
+                            } else {
+                                attachedBlocks.put(adjacentPos, adjacentBlockId);
+                                blocksChecked.add(adjacentPos);
+                                blocksToCheck.add(adjacentPos.add(movement));
+                            }
+                        }
+                    }
+                }
+                // Check next block in line
+                blocksToCheck.add(blockPos.add(movement));
+            } else if (!BlockStateValues.canPistonDestroyBlock(blockId)) {
+                // Block can't be moved or destroyed, so it blocks all block movement
+                moveBlocks = false;
+                break;
+            }
+        }
+        if (!moveBlocks || attachedBlocks.size() > 12) {
+            attachedBlocks.clear();
+        } else {
+            flattenPositions();
+        }
+    }
+
+    private boolean canMoveBlock(int javaId, boolean isPushing) {
+        if (javaId == BlockStateValues.JAVA_AIR_ID) {
+            return true;
+        }
+        // Pistons can only be moved if they aren't extended
+        if (PistonBlockEntityTranslator.isBlock(javaId)) {
+            return !BlockStateValues.getPistonValues().get(javaId);
+        }
+        BlockMapping block = BlockRegistries.JAVA_BLOCKS.getOrDefault(javaId, BlockMapping.AIR);
+        // Bedrock, End portal frames, etc. can't be moved
+        if (block.getHardness() == -1.0d) {
+            return false;
+        }
+        switch (block.getPistonBehavior()) {
+            case BLOCK:
+            case DESTROY:
+                return false;
+            case PUSH_ONLY: // Glazed terracotta can only be pushed
+                return isPushing;
+        }
+        // Pistons can't move block entities
+        return !block.isBlockEntity();
+    }
+
+    /**
+     * Get the unit vector for the direction of movement
+     *
+     * @return The movement of the blocks
+     */
+    private Vector3i getMovement() {
+        if (action == PistonValueType.PULLING) {
+            return orientation.reversed().getUnitVector();
+        }
+        return orientation.getUnitVector(); // PUSHING and CANCELLED_MID_PUSH
+    }
+
+    /**
+     * Replace all attached blocks with air
+     */
+    private void removeBlocks() {
+        for (Vector3i blockPos : attachedBlocks.keySet()) {
+            ChunkUtils.updateBlock(session, BlockStateValues.JAVA_AIR_ID, blockPos);
```

Comment:
If this method is only called for visuals, I would rather it just call the update block methods manually so we don't even touch the chunk cache.